warehouse/accounts/services.py

```diff
@@ -621,6 +626,79 @@ def get_password_timestamp(self, user_id):
         user = self.get_user(user_id)
         return user.password_date.timestamp() if user.password_date is not None else 0
 
+    def check_device_valid(self, user_id: str, device_id_secret_b64: str) -> bool:
+        """
+        Checks whether the given device id and secret are valid for the given user.
+        """
+        self._metrics.increment("warehouse.authentication.check_device_valid.start")
+
+        if not device_id_secret_b64:
+            return False
+        try:
+            device_id_secret = DeviceIdSecret.from_base64(device_id_secret_b64)
+        except (KeyError, ValueError):
+            return False
+
+        self._check_ratelimits(
+            userid=user_id,
+            tags=["mechanism:check_device_valid"],
+        )
+
+        user = self.get_user(user_id)
+        for device in user.user_devices:
+            if device.device_id != device_id_secret.device_id:
+                continue
+
+            is_expired = (
+                datetime.datetime.now() - device.saved_date
+                > datetime.timedelta(days=SAVED_DEVICE_VALID_DAYS)
+            )
+            if is_expired:
+                continue
+
+            if self.hasher.verify(device_id_secret.secret, device.device_secret):
+                self._metrics.increment(
+                    "warehouse.authentication.check_device_valid.ok"
+                )
+                return True
+
+        self._metrics.increment("warehouse.authentication.check_device_valid.failure")
+        return False
+
+    def generate_device_id_secret(self, user_id: str) -> str:
+        """
+        Generates and saves to the database a new device id and secret for the given
+        user. Returns the device id and secret as a base64 encoded string.
+        """
+        user = self.get_user(user_id)
+
+        device_id = self._generate_device_id(user)
+        device_secret = secrets.token_urlsafe(DEVICE_SECRET_BYTES)
+
+        self.db.add(
+            UserDevice(  # type: ignore
+                user=user,
+                device_id=device_id,
+                device_secret=self.hasher.hash(device_secret),
+            )
+        )
+        self.db.flush()
+
+        return DeviceIdSecret(device_id, device_secret).to_base64()
+
+    def _generate_device_id(self, user: User) -> str:
+        """
+        Generates a unique device id for the given user.
+        """
+        attempts_max = 1000
+        attempts = 0
+        while attempts < attempts_max:
+            device_id = secrets.token_urlsafe(DEVICE_ID_BYTES)
+            if not any(device.device_id == device_id for device in user.user_devices):
+                return device_id
+            attempts += 1
+        raise RuntimeError("Failed to generate device id")
```

Comment:
Can you explain why we need a device ID _and_ a device secret? My first thought is that it should be sufficient to generate a random token, store it in a cookie, and store the hash of that token along with a reference to the user in the database.

Reply:
We could even just model it as

```
class UserDevice(db.Model):
     __tablename__ = "user_devices"

     user_id = Column(
         UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
     )
     device_id = Column(
          UUID(as_uuid=True), unique=True, nullable=False, index=True
     )
     saved_date = Column(
         DateTime(timezone=False), nullable=False, server_default=sql.func.now()
     )
     device_info = # ....maybe a JSONB column?
```

And instead of storing a secret at all, use the `TokenService` to send a signed token containing `{"username": User.username, "device_id": UserDevice.id}`.

We can then just see if the signed token validates.

Reply:
Yes, I thought it would be important to use a slow hash with a salt for the device secret, so that if an attacker gained access to the database, they would have a very hard time reversing the hash and finding the original secret. Using a salt means we have to hash the secret for each value we want to compare in the database. Since we're using a slow hash, I didn't want to have to hash the secret multiple times; if a user had several devices, that could add noticeable time. This is why I introduced the device ID. 

Now, I'll admit, I might be going a bit overkill with the security here; a slow hash with a salt might not be necessary with a randomly-generated 128-bit string. It could be fine storing a plain SHA256 hash of the secret in the database. But since this is a highly security sensitive area, I figured I would rather err on the side of too much security.

In regards to sending a signed token instead, if we could store all the necessary information in the token and not need a new database table, I think that would be the better option. However, since we will probably want to give the user the ability to revoke devices, I think a database table is necessary. I think using a signed token or a hash of a secret are both equally good approaches. I'm inclined to leave it as a hash of a secret since that's what I already have, but if there are advantages to the signed token that I'm not thinking of, I am willing to switch to that.

Reply:
I agree keeping a record in the DB for the Device is necessary for revocation. Only reason I think I prefer the token approach is that it makes the data in the table a bit more transparent, I'm not committed to it. Definitely welcome other thoughts.

Reply:
I heard you on the brute-force side of things, but I'm still leaning towards using a TokenService rather than storing the hashed secret in the database.

Not to rely on tokens themselves, but just for the handling. When a device is saved we still retain the data in the DB for revocation purposes, we just send a signed token to the client that they can provide at login for warehouse to validate.

Any thoughts from the other @pypi/warehouse-admins?

Reply:
I can change this to use a TokenService instead. I'm thinking that to complete this issue (https://github.com/pypi/warehouse/issues/5867), we won't need to add the database table yet. When we do the followup work to add the revocation, we can add the database table then. For now, we would just store the cookie in the user's browser and then verify it on the server side. Does that sound reasonable?

Reply:
@ewdurbin What are your thoughts on my previous question?

Reply:
I'll plan on using a signed token that is verified server side without a database table at the current time. I'm going to close this PR and create a new one since this is a significant rework.